source/k64f_eth_nanostack_port.c

```diff
@@ -501,236 +361,275 @@ static int8_t arm_eth_phy_k64f_address_write(phy_address_type_e address_type, ui
 
-    else if (link_status && last_known_status==LINK_STATUS_DOWN) {
-            last_known_status = LINK_STATUS_UP;
-            ethernet_iface_ptr->isInitialized = true;
-            tr_debug("Ethernet phy link status = up");
-            tr_info("Ethernet Cable plugged.");
-            enet_phy_link_setup(ethernet_iface_ptr);
+    /* Preparing the Buffer Configurations */
+    enet_buffer_config_t buf_config = { .rxBdNumber = ENET_RX_RING_LEN,
+        .txBdNumber = ENET_TX_RING_LEN, .rxBuffSizeAlign = ENET_ALIGN(
+                ENET_ETH_MAX_FLEN, ENET_BUFF_ALIGNMENT), .txBuffSizeAlign = 0,
+        .rxBdStartAddrAlign = base_DS.rx_buf_desc_start_addr,
+        .txBdStartAddrAlign = base_DS.tx_buf_desc_start_addr,
+        // it appears this is a wrongly-typed pointer to an array of buffer pointers
+        .rxBufferAlign = (uint8_t *) base_DS.rx_buf, .txBufferAlign = NULL };
+
+    /* Initialize the Ethernet Hardware */
+    initialize_enet_hardware();
+
+    /* Setup Clock for ENET module */
+    sysClock = CLOCK_GetFreq(kCLOCK_CoreSysClk);
+
+    /* Load the Default config (RMII Mode, Full Duplex, 100 Mbps)*/
+    ENET_GetDefaultConfig(&config);
+
+    /* Initialize PHY layer */
+    PHY_Init(ENET, phyAddr, sysClock);
+
+    /* Setup PHY layer */
+    PHY_GetLinkStatus(ENET, phyAddr, &link_currently_up);
+    tr_info("Ethernet cable is %sconnected.", link_currently_up ? "" : "not ");
+    if (link_currently_up) {
+        /* Get link information from PHY (Auto negotiation, in case deafult
+         * config is different from actual link available )  */
+        PHY_GetLinkSpeedDuplex(ENET, phyAddr, &phy_speed, &phy_duplex);
+        /* Change the MII speed and duplex for actual link status. */
+        config.miiSpeed = (enet_mii_speed_t) phy_speed;
+        config.miiDuplex = (enet_mii_duplex_t) phy_duplex;
+        config.interrupt = kENET_RxFrameInterrupt | kENET_TxFrameInterrupt;
     }
+    config.rxMaxFrameLen = ENET_ETH_MAX_FLEN;
+    config.macSpecialConfig = kENET_ControlFlowControlEnable;
+    config.rxAccelerConfig = kENET_RxAccelMacCheckEnabled;
+
+    /* Initialize ENET module */
+    ENET_Init(ENET, &global_enet_handle, &config, &buf_config, mac_addr,
+              sysClock);
+
+    /* Adding Multicast Group
+     * Even ksdk 2.0 API is weird.
+     * ENET_AddMulticastGroup(ENET_Type *base, uint8_t *address),
+     * this routine will need 64 addresses to set a multicast group.
+     * So let's hack it. For more details->Kevin*/
+    ENET->GAUR = 0xFFFFFFFFu;
+    ENET->GALR = 0xFFFFFFFFu;
+
+    /* Setup callback for Interrupt routines */
+    ENET_SetCallback(&global_enet_handle, ethernet_event_callback, NULL);
+
+    /* Lastly, activate the module */
+    ENET_ActiveRead(ENET);
 }
 
-/* Interrupt handling Section*/
 
-/* Enables Ethernet interrupts */
-void eth_enable_interrupts(void)
+/** \brief  Function to get a lock on the Thread.
+ * An abstraction of platform_enter_critical. Makes sure that we have the mutex.
```

Comment:
Description isn't very accurate.

In RTOS mode, it claims a mutex, protecting foreground operations against the background tasks enet_rx/tx_task (because they themselves are called inside the lock).

In non-RTOS mode, it disables interrupts, protecting foreground operations against the background interrupts enet_rx/tx_task.